src/cmap/auth/scram.ts

```diff
@@ -109,13 +122,19 @@ function executeScram(cryptoMethod: any, authContext: any, callback: Function) {
 }
 
 function continueScramConversation(
-  cryptoMethod: any,
-  response: any,
-  authContext: any,
-  callback: Function
+  cryptoMethod: CryptoMethod,
+  response: Document,
+  authContext: AuthContext,
+  callback: Callback
 ) {
   const connection = authContext.connection;
   const credentials = authContext.credentials;
+  if (!credentials) {
+    return callback(new MongoError('AuthContext must provide credentials.'));
```

Comment:
This made me wonder: would we ever be executing an auth mechanism if credentials were not provided? We only call `auth` in `connect` if credentials are detected, so I think it should not be an optional property on `AuthContext`

Reply:
This one is tough because of the way we set up our speculative auth, the prepare side of auth takes an AuthContext since it will need to tack on properties that get started in that phase, but at that time the credentials don't have to be specified so I made credentials on AuthContext optional which means threading through this if statement to each authenticator. The alternative is an if statement over in connect.ts that skips the prepare call, but code structure being the way it is, callbacks and all, I think I'd have to do a larger refactor to break out the callbacks into named functions for reusability or big copy pastas.

I could missing something here though, does that logic track?